# Part 1: Data Loading and Exploration

In [39]:
import pandas as pd
from sklearn.datasets import fetch_california_housing

# Load the housing dataset
housing = fetch_california_housing()

# Create a data frame
X = pd.DataFrame(housing.data, columns=housing.feature_names) 
y = pd.Series(housing.target, name='med_house_value')

# Initial exploration
# Show feature variables 
all_vars = ["MedInc", "HouseAge", "AveRooms", "AveBedrms", "Population", "AveOccup", "Lattitude", "Longitude"]
print(all_vars)
print(X.isnull().sum())
print(X.head())
X.describe()




['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Lattitude', 'Longitude']
MedInc        0
HouseAge      0
AveRooms      0
AveBedrms     0
Population    0
AveOccup      0
Latitude      0
Longitude     0
dtype: int64
   MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  
0    -122.23  
1    -122.22  
2    -122.24  
3    -122.25  
4    -122.25  


,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude
count,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000
mean,3.870671,28.639486,5.429000,1.096675,1425.476744,3.070655,35.631861,-119.569704
std,1.899822,12.585558,2.474173,0.473911,1132.462122,10.386050,2.135952,2.003532
min,0.499900,1.000000,0.846154,0.333333,3.000000,0.692308,32.540000,-124.350000
25%,2.563400,18.000000,4.440716,1.006079,787.000000,2.429741,33.930000,-121.800000
50%,3.534800,29.000000,5.229129,1.048780,1166.000000,2.818116,34.260000,-118.490000
75%,4.743250,37.000000,6.052381,1.099526,1725.000000,3.282261,37.710000,-118.010000
max,15.000100,52.000000,141.909091,34.066667,35682.000000,1243.333333,41.950000,-114.310000


# Part 2: Linear Regression on Unscaled Data

In [35]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

# Split the raw data (80% training, 20% testing)
X_train_raw, X_test_raw, y_train, y_test,  = train_test_split(X, y, test_size=0.2, random_state= 42)
# Initialize and train the linear regression model on unscaled data
lin_reg_raw = LinearRegression()
lin_reg_raw.fit(X_train_raw, y_train)
y_pred_raw = lin_reg_raw.predict(X_test_raw)



In [36]:
#Model performance evaluation 
from sklearn.metrics import mean_squared_error, root_mean_squared_error, r2_score
mse_raw = mean_squared_error( y_test, y_pred_raw)
rmse_raw = root_mean_squared_error(y_test, y_pred_raw)
r2_raw = r2_score(y_test, y_pred_raw)

# Showing the metrics
print("Unscaled Data Model:")
print(f"Mean squared error: {mse_raw:.2f}")
print(f"Root mean squared error: {rmse_raw:.2f}")
print(f"R² Score: {r2_raw:.2f}")




Unscaled Data Model:
Mean squared error: 0.56
Root mean squared error: 0.75
R² Score: 0.58


In [37]:
# Coefficients & intercept
Coef_series= pd.Series(lin_reg_raw.coef_, index=X.columns)
print("Model Coefficents (Unscaled):")
print(Coef_series)
 
intercept = lin_reg_raw.intercept_
print(f"Intercept:", intercept)

Model Coefficents (Unscaled):
MedInc        0.448675
HouseAge      0.009724
AveRooms     -0.123323
AveBedrms     0.783145
Population   -0.000002
AveOccup     -0.003526
Latitude     -0.419792
Longitude    -0.433708
dtype: float64
Intercept: -37.0232777060643


#### What does the R² score tell us about model performance?
Interpretation: The R² score represents the fit of the of the model and how much variance is captured. An R²  of 1 is a perfect fit while 0 is model failing to capture variance. The R²  here is .58 of variance is explained, overall the score is closer to 1 than 0 and greater than .5 which gives us some reassurance that the model is fine but could be better. 

#### Which features seem to have the strongest impact on predictions based on the model’s coefficients?

The model coefficients here with the strongest impact are first Average Bedrooms which have a postive impact of .783. In unscaled coefficients this is interpreted as a 1 unit increase in aveBedrms would increase median house value by .783 assuming all other factors are heald constant. The next strongest coefficient is medium income with a positive coefficient of .449. Right after that is Longitude, - .434 and Latitude -.42 which both have a negative effect on house value. 

#### How well do the predicted values match the actual values?
 The mean squared error which is .56 measures the average squared difference between actual and predicted house values. RMSE is the square root of MSE, gives an error metric in the same units as the target which here is .75 a lower rmse means better predictive performance so there is room for improvement here. Overall there is a mismatch between predicted values and actual that should be addressed to make the model more of a better fit. 

# Part 3: Feature Selection and Simplified Model

In [72]:
feature_vars = ["AveBedrms","MedInc", "AveRooms"]
X_simplified = pd.DataFrame(X[feature_vars])
#Split data 
X_train__simplified, X_test__simplified, y_train_simplified, y_test_simplified,  = train_test_split(X_simplified, y, test_size=0.2, random_state= 42)
#Initialize and Train data 
lin_reg_simplified = LinearRegression()
lin_reg_simplified.fit(X_train__simplified,y_train_simplified)
# Make predictions on the simplified test set
y_pred_simplified = lin_reg_simplified.predict(X_test__simplified)
# Evaluate model performance
mse_simplified = mean_squared_error(y_test_simplified, y_pred_simplified)
rmse_simplified = root_mean_squared_error(y_test_simplified, y_pred_simplified)
r2_simplified = r2_score(y_test_simplified, y_pred_simplified)

In [73]:
# Showing the metrics
print("Simplified Data Model:")
print(f"Mean squared error: {mse_simplified:.2f}")
print(f"Root mean squared error: {rmse_simplified:.2f}")
print(f"R² Score: {r2_simplified:.2f}")

Simplified Data Model:
Mean squared error: 0.68
Root mean squared error: 0.82
R² Score: 0.48


#### How does the simplified model compare to the full model?
The simplified model is even less of a fit then the full model. I think this could be because of ommited variable bias such as the fact Latitude and Longitude have stronger negative coefficients and not included in the simplified model.  The R² score is lower by .10 meaning less variation is explained by the simplified model. The mean squared error and root mean squared error are higher meaning that the predicted values are less of a fit for the actual values since we prefer lower rmse values. 
#### Would you use this simplified model in practice? Why or why not?
I would not be in favor of using the simplified model because we loose more variation and create a higher gap between predictive and actual values, the whole model would give us more of an explanation. The simplified model can be used when we want to examine specific feature impact so we are not looking at everything all at once. 